ublox/src/ubx_packets/packets.rs

```diff
@@ -0,0 +1,718 @@
+    itow: u32,
+    time_accuracy_estimate_ns: u32,
+    /// Nanoseconds of second, range -1e9 .. 1e9
+    nanos: i32,
+    /// Year, range 1999..2099
+    year: u16,
+    /// Month, range 1..12
+    month: u8,
+    /// Day of Month, range 1..31
+    day: u8,
+    /// Hour of Day, range 0..23
+    hour: u8,
+    /// Minute of Hour, range 0..59
+    min: u8,
+    /// Seconds of Minute, range 0..59
+    sec: u8,
+    /// Validity Flags
+    #[ubx(map_type = NavTimeUtcFlags)]
+    valid: u8,
+}
+
+#[ubx_extend_bitflags]
+#[ubx(from, rest_reserved)]
+bitflags! {
+    /// Validity Flags of `NavTimeUTC`
+    pub struct NavTimeUtcFlags: u8 {
+        /// Valid Time of Week
+        const VALID_TOW = 1;
+        /// Valid Week Number
+        const VALID_WKN = 2;
+        /// Valid UTC (Leap Seconds already known)
+        const VALID_UTC = 4;
+    }
+}
+
+/// Navigation/Measurement Rate Settings
+#[ubx_packet_send]
+#[ubx(class = 6, id = 8, fixed_payload_len = 6)]
+struct CfgRate {
+    /// Measurement Rate, GPS measurements are taken every `measure_rate_ms` milliseconds
+    measure_rate_ms: u16,
+    /// Navigation Rate, in number of measurement cycles.
+    /// On u-blox 5 and u-blox 6, this parametercannot be changed, and is always equals 1.
+    nav_rate: u16,
+    /// Alignment to reference time
+    #[ubx(map_type = AlignmentToReferenceTime)]
+    time_ref: u16,
+}
+
+/// Alignment to reference time
+#[repr(u16)]
+#[derive(Clone, Copy, Debug)]
+pub enum AlignmentToReferenceTime {
+    Utc = 0,
+    Gps = 1,
+}
+
+impl AlignmentToReferenceTime {
+    const fn into_raw(self) -> u16 {
+        self as u16
+    }
+}
+
+/// Set Message Rate the current port
+#[ubx_packet_send]
+#[ubx(class = 6, id = 1, fixed_payload_len = 3)]
+struct CfgMsg3 {
+    msg_class: u8,
+    msg_id: u8,
+    /// Send rate on current Target
+    ///
+    rate: u8,
+}
+
+impl CfgMsg3Builder {
+    #[inline]
+    pub fn set_rate_for<T: UbxPacketMeta>(rate: u8) -> Self {
+        Self {
+            msg_class: T::CLASS,
+            msg_id: T::ID,
+            rate,
+        }
+    }
+}
+
+/// Set Message rate configuration
+/// Send rate is relative to the event a message is registered on.
+/// For example, ifthe rate of a navigation message is set to 2,
+/// the message is sent every secondnavigation solution
+#[ubx_packet_send]
+#[ubx(class = 6, id = 1, fixed_payload_len = 8)]
+struct CfgMsg8 {
+    msg_class: u8,
+    msg_id: u8,
+    /// Send rate on I/O Port (6 Ports)
+    rates: [u8; 6],
+}
+
+impl CfgMsg8Builder {
+    #[inline]
+    pub fn set_rate_for<T: UbxPacketMeta>(rates: [u8; 6]) -> Self {
+        Self {
+            msg_class: T::CLASS,
+            msg_id: T::ID,
+            rates,
+        }
+    }
+}
+
+/// Receiver/Software Version
+#[ubx_packet_recv]
+#[ubx(class = 0x0a, id = 0x04)]
+struct MonVer {
+    #[ubx(map_type = &str, may_failed, from = mon_ver::convert_to_str_unchecked,
+          is_valid = mon_ver::is_cstr_valid, get_as_ref)]
```

Comment:
- `is_valid` used to distinguish packets with the same class and the same message id,
    like CfgPrt(Uart|Spi|Usb)
- it will be very inconvenient to use it, imagine you got PacketX with several invalid fields,
   You need manually check all fields like:
       ```rust
       match (pack_ref.field1(), pack_ref.field2(), pack_ref.field3()) {
          (Some(f1), Some(f2), Some(f3)) => ok_we_got_what_we_want,
          None => return err,
       }
      ```
      and `Option` means that there is no such field, while there is such field, but packet broken, and why user have to manually handle such errors?

In case of MonVer, missed 0 char, extenstion length not match 30 * N, means that something wrong with channel, or something like bit-flip happens in memory of GPS receiver, so all data inside of packet may be not valid, TCP checksum used in UBX protocol is not that [great](https://stackoverflow.com/questions/3830206/can-a-tcp-checksum-fail-to-detect-an-error-if-yes-how-is-this-dealt-with)